### PR TITLE
Add parsing for 8_367_33.

### PR DESCRIPTION
**Pull request by andyvan-trabus**
This was a lot more work, thanks for the suggestion to work my way up to it.

I did make 1 change that I think you *might* push back against.  I ran cppcheck on the new files, and I was getting warnings about missing assignments in the base AisMsg class.  I added them in to eliminate the warnings in the new files.

### CODE REVIEW

File: src/libais/ais.h [hunk 1]
```diff
@@ -2037,6 +2037,318 @@ class Ais8_367_25 : public Ais8 {
...
+ public:
+  int speed_raw = 0;  // 0.1 Knots
+  float speed = 0.0;  // Knots
+  int dir = 0;  // Degrees
+  int depth = 0;  // Meters
+};
+
+class Ais8_367_33_Curr2D : public Ais8_367_33_SensorReport {
+ public:
+  Ais8_367_33_Curr2D_Current currents[3];
```
Comment (schwehr) [inline src/libais/ais.h:2158]: Prefer std::array here and everywhere else

File: src/libais/ais.h [hunk 2]
```diff
@@ -2037,6 +2037,318 @@ class Ais8_367_25 : public Ais8 {
...
+  int spare2 = 0;
+
+  Ais8_367_33_Location(const AisBitset &bs, const size_t offset);
+  ~Ais8_367_33_Location() {}
+  Ais8_367_33_SensorEnum getType() const override { return AIS8_367_33_SENSOR_LOCATION; }
+};
+
+class Ais8_367_33_Station : public Ais8_367_33_SensorReport {
+ public:
+  string name = "";
```
Comment (schwehr) [inline src/libais/ais.h:2099]: std::string.  And no need to set to "".

**Review by schwehr — CHANGES_REQUESTED** [review]

**Comment by schwehr** [comment]
> This was a lot more work, thanks for the suggestion to work my way up to it.
> 
> I did make 1 change that I think you _might_ push back against. I ran cppcheck on the new files, and I was getting warnings about missing assignments in the base AisMsg class. I added them in to eliminate the warnings in the new files.

Those cppcheck fixes are correct to initialize ints and floats.

File: src/libais/ais.h (hunk 1 above)
Comment (andyvan-trabus) [reply src/libais/ais.h:2158]: Okay.  I was modeling my changes after AIS8_1_26.  Will make the change.

File: src/libais/ais.h (hunk 2 above)
Comment (andyvan-trabus) [reply src/libais/ais.h]: Done

File: src/libais/ais8_367.cpp [hunk 3]
```diff
@@ -274,5 +274,459 @@ ostream& operator<< (ostream &o, const Ais8_367_25 &msg) {
...
+  timeout = bits.ToUnsignedInt(offset + 80, 3);
+  spare2 = bits.ToUnsignedInt(offset + 83, 2);
+}
+
+Ais8_367_33_Station::Ais8_367_33_Station(const AisBitset &bits,
+                                         const size_t offset) {
+  name = bits.ToString(offset, 84);
+  // Remove any trailing '@' characters.  Spec says they are not to be shown in the presentation.
+  while (name.back() == '@') {
+    name.pop_back();
```
Comment (schwehr) [inline src/libais/ais8_367.cpp:298]: https://www.cplusplus.com/reference/string/string/back/ says:

`This function shall not be called on empty strings.`

While it currently works with clang and gcc there needs to be an empty check first.

```
while (!name.empty() && name.back() == '@') {
```

https://godbolt.org/z/1cPYoM8GT

File: src/libais/ais8_367.cpp [hunk 4]
```diff
@@ -274,5 +274,459 @@ ostream& operator<< (ostream &o, const Ais8_367_25 &msg) {
...
+  utc_hour_forecast = bits.ToUnsignedInt(offset + 49, 5);
+  utc_min_forecast = bits.ToUnsignedInt(offset + 54, 6);
+  duration = bits.ToUnsignedInt(offset + 60, 8);
+  spare2 = bits.ToUnsignedInt(offset + 68, 17);
+}
+
+Ais8_367_33_Curr2D::Ais8_367_33_Curr2D(const AisBitset &bits,
+                                       const size_t offset) {
+  for (size_t idx = 0; idx < 3; idx++) {
+    size_t start = offset + idx * 26;
```
Comment (schwehr) [inline src/libais/ais8_367.cpp:339]: `const size_t start`  Better to be aggressive with adding const here and elsewhere.  It doesn't really change anything for the compiler, but it tells the reader that there aren't going to be any hidden surprises like `++`

File: src/libais/ais8_367.cpp [hunk 5]
```diff
@@ -274,5 +274,459 @@ ostream& operator<< (ostream &o, const Ais8_367_25 &msg) {
...
+
+  if (!CheckStatus()) {
+    return;
+  }
+  if (num_bits  < 168 || num_bits > 952) {
+    status = AIS_ERR_BAD_BIT_COUNT;
+    return;
+  }
+
+	const size_t report_start = 56;
```
Comment (schwehr) [inline src/libais/ais8_367.cpp:580]: Funky indenting

File: src/libais/ais8_367.cpp [hunk 6]
```diff
@@ -274,5 +274,459 @@ ostream& operator<< (ostream &o, const Ais8_367_25 &msg) {
...
+      status = AIS_ERR_BAD_SUB_MSG;
+      return;
+    }
+  }
+
+  status = AIS_OK;
+}
+
+ostream& operator<< (ostream &o, const Ais8_367_33 &msg) {
+	int num_reports = msg.reports.size();
```
Comment (schwehr) [inline src/libais/ais8_367.cpp:606]: `const int num_reports`

File: src/libais/ais8_367.cpp [hunk 7]
```diff
@@ -274,5 +274,459 @@ ostream& operator<< (ostream &o, const Ais8_367_25 &msg) {
...
+	for (int report_idx = 0; report_idx < num_reports; report_idx++) {
```
Comment (schwehr) [inline src/libais/ais8_367.cpp:607]: Please indent by 2 spaces

File: src/libais/ais_py.cpp [hunk 8]
```diff
@@ -2284,9 +2881,12 @@ ais8_to_pydict(const char *nmea_payload, const size_t pad) {
     case AIS_FI_8_367_24_SSW_SMALL:
       status = ais8_367_24_append_pydict(nmea_payload, dict, pad);
       break;
-  case AIS_FI_8_367_25_SSW_TINY:
+    case AIS_FI_8_367_25_SSW_TINY:
       status = ais8_367_25_append_pydict(nmea_payload, dict, pad);
       break;
+  case AIS_FI_8_367_33_ENVIRONMENTAL:
```
Comment (schwehr) [inline src/libais/ais_py.cpp:2887]: Match indent

**Review by schwehr — APPROVED** [review]

**Pull request merged**